### ISSUE TITLE
Guidance for EBI/EMBL submission to ENA/SRA

### ISSUE THREAD

**Issue by peterjc** [tseemann/prokka#145]
The main README file has a short guide for submission to the NCBI GenBank, but it would be nice to have a similar section for submission to the ENA/SRA via the EBI/EMBL.

Pre-submission annotation validation should be done, see http://www.ebi.ac.uk/ena/software/flat-file-validator for details. I presume they can take the GenBank output from Prokka just as easily as an EMBL flat file, since both have a valid INSDC feature table? *(Not the case, use EMBL format - see below)*

```
$ curl -L -O ftp://ftp.ebi.ac.uk/pub/databases/ena/lib/embl-client.jar
$ java -jar embl-client.jar -r -f genbank mydir/mygenome.gbk
```

As with GenBank submission, I suspect you should preregister the project to get a project ID and declare you locus ID prefix.

(Right now I am struggling though this, see e.g. #47 which is solved.)

**Comment by andrewjpage** [comment]
https://github.com/sanger-pathogens/gff3toembl

**Comment by peterjc** [comment]
Thanks @andrewjpage - I take it that means the GenBank files from Prokka do not work as is, since you went to the trouble of generating EMBL files from Prokka's (GFF) output.

**Comment by andrewjpage** [comment]
Theres a few edge cases and undocumented features when submitting to EMBL.  If @tsseman was to try and keep up with every change to their API (which they usually dont announce), he would spend his life updating Prokka!

**Comment by peterjc** [comment]
I emailed to report a bug in the GenBank validation failing to understand the ``DBLINK`` line which I added to the file to report the project number (giving ``ERROR: Unknown line type: DBLINK      BioProject: PRJ...``).

After a few emails with the ENA Data Submission team, it looks like EMBL format is the way to go - even though the current validator says it supports GenBank and GFF3 formats via the ``-f`` option:

```
$ java -jar embl-client.jar -help
Usage: ena_validator <files> [options]
 
  Options:    -assembly        genome assembly entries (default: false)
    -db              Database connection required for some checks and fixes.Valid databases : ENAPRO or ENADEV. Can only be used within EBI. (optional)(default :false)
    -de              Additional Fix :Adds/Fixes DE line(optional) (default: false)
    -f               File format(optional) Values:'embl','genbank','gff3','assembly' (default: embl)
    -filter          -filter <prefix> Store entries in <prefix>_good.txt and <prefix>_bad.txt files in the working directory. Entries with errors are stored in the bad file and entries without errors are stored in the good file. (optional)(default :false)
    -fix             Fixes entries in input files. Stores input files in 'original_files' folder. (optional) (default: false)
    -fix_diagnose    Creates 'diagnose' folder in the current directory with original entries in <filename>_origin file and the fixed entries in <filename>_fixed file. Only fixed entries will be stored in these files.(optional)  (default: false)
    -help            Displays available options (default: false)
    -l               Log level(optional) Values : 0(Quite),1(Summary),2(Verbose) (default: 1)
    -lowmemory       Runs in low memory usage mode. Writes error logs but does not show message summary(optional) (default: false)
    -min_gap_length  minimum gap length to generate assembly_gap/gap features, use assembly flag to add assembly_gap features (default: 0)
    -prefix          Adds prefix to report files
    -r               Remote, is this being run outside the EBI(optional) (default: false)
    -skip            -skip <errorcode1>,<errorcode2>,... Ignore specified errors.(optional)(default:false) 
    -wrap            Turns on line wrapping in flat file writing (optional)  (default: false)
```

I was told:

*Yes, we advise to use the validator in EMBL flat file format only.
I will ask to check the documentation since only this should be stated.*

I'll look at the Sanger Pathogen groups' gff3toembl tool next - thanks for thank link @andrewjpage

**Comment by peterjc** [comment]
Something like this looks like a plan:

* Register your project and locus tag on the ENA (make a note of the taxid for later)

* Install Python 2.7 (currently required for conversion tool)

* Install genometools from https://github.com/genometools/genometools (don't need Cairo, so use ``make cairo=no ...`` if you haven't got Cairo installed)

* Install gff3toembl from https://github.com/sanger-pathogens/gff3toembl

* Run the converter, filling in the title, if circular (lucky you, PacBio?) or linear, species name, taxid, etc.

```
gff3_to_embl -i "Submitter, A." \
-m "Bacullis sp. strain XYZ genome annotated using Prokka." \
-g linear -c PROK -n 11 -f submission.embl \
"Bacullis sp." 1386 PRJEB12345 "Bacullis sp. strain XYZ" mydir/from_prokka.gff
```

This will use the ``-i`` author information for the ``RA`` lines (use surname space initial dot initial dot comma space etc, apparently just the submitter is enough), the ``-m`` title for the ``RT`` lines. If giving ``-p`` (publication) this goes to the ``RL`` line, but that will safely default to something the validator accepts: ``Submitted (DD-MMM-YYYY) to the INSDC.`` with today's date.

* Download the ENA validator, see http://www.ebi.ac.uk/ena/software/flat-file-validator

```
$ curl -L -O ftp://ftp.ebi.ac.uk/pub/databases/ena/lib/embl-client.jar
```

* Run the validator,

```
$ java -jar embl-client.jar -r submission.embl
```

If you get error messages with the cryptic string ``{0}`` this is probably a bug (which I reported), try looking in the ``VAL_ERROR.txt`` etc files which seem more complete.

**Comment by aleimba** [comment]
Hi @peterjc,
looks like a good plan, thanks for following up. One question, are you correcting the errors from the validator directly in the EMBL text file?

**Comment by andrewjpage** [comment]
The first step can be simplified slightly. We don't register a locus tag at all, we use the sample accession number for the locus tag by convention, and EMBL magically make it work on their end.

**Comment by peterjc** [comment]
@andrewjpage could you clarify what you mean about the locus tag - do you register the project/sample with ENA, and then use the sample accession as the locus tag argument to run prokka? (In this case I can't do that - we'd already run prokka and assigned locus tags before starting the ENA submission)

@aleimba Touch wood I have been able to get the validation to pass without any manual edits of the Prokka output, just careful choice of options to ``gff3_to_embl``. For example, trying to set the ``-p`` option gave what looked to me like a sensible ``RL`` line but the validator complained about it. I hope to do the submission later today...

**Comment by andrewjpage** [comment]
@peterjc When we run gff3_to_embl we pass in the sample accession number
(ERS1234) into the --locus_tag parameter and it updates all the locus tags
in the annotation.   Then we submit via the API.  We dont register the
locus tag in advance, but we do register the project/sample in advance.

**Comment by dutchscientist** [comment]
Thanks, this is great, it helped me to solve https://github.com/sanger-pathogens/gff3toembl/issues/38 :)

**Comment by peterjc** [comment]
See also #113 which discusses generating EMBL format directly within Prokka.

**Comment by peterjc** [comment]
I submitted pull request #148 a while back to document this approach in the main README file.

**Comment by peterjc** [comment]
Closed via pull request #148, thanks for merging this Torsten.
